core/trino-main/src/main/java/io/trino/operator/aggregation/partial/PartialAggregationOutputProcessor.java

```diff
@@ -0,0 +1,123 @@
+    }
+
+    private PartialAggregationOutputProcessor(List<Integer> groupByChannels, Optional<Integer> inputHashChannel, List<AggregatorFactory> aggregatorFactories, List<? extends Type> aggregationInputTypes, List<Integer> aggregationInputChannels, List<Integer> maskBlockChannels)
+    {
+        this.aggregationInputTypes = ImmutableList.copyOf(aggregationInputTypes);
+
+        this.groupedAggregators = requireNonNull(aggregatorFactories, "aggregatorFactories is null").stream().map(AggregatorFactory::createGroupedAggregator).collect(toImmutableList());
+        this.aggregationInputChannels = Ints.toArray(aggregationInputChannels);
+        this.maskBlockChannels = Ints.toArray(maskBlockChannels);
+        this.hashChannels = new int[groupByChannels.size() + (inputHashChannel.isPresent() ? 1 : 0)];
+        for (int i = 0; i < groupByChannels.size(); i++) {
+            hashChannels[i] = groupByChannels.get(i);
+        }
+        inputHashChannel.ifPresent(channelIndex -> hashChannels[groupByChannels.size()] = channelIndex);
+        finalBlockCount = hashChannels.length + groupedAggregators.size() + maskBlockChannels.size() + aggregationInputChannels.size() + 1;
+    }
+
+    public Page processAggregatedPage(Page page)
+    {
+        Block[] finalPage = new Block[finalBlockCount];
+        int blockOffset = 0;
+        for (int i = 0; i < page.getChannelCount(); i++, blockOffset++) {
+            finalPage[blockOffset] = page.getBlock(i);
+        }
+        int positionCount = page.getPositionCount();
+
+        // mask channels
+        for (int i = 0; i < maskBlockChannels.length; i++, blockOffset++) {
+            finalPage[blockOffset] = RunLengthEncodedBlock.create(groupedAggregators.get(i).getType(), null, positionCount);
+        }
+        // aggregation inputs
+        for (int i = 0; i < aggregationInputTypes.size(); i++, blockOffset++) {
+            finalPage[blockOffset] = RunLengthEncodedBlock.create(aggregationInputTypes.get(i), null, positionCount);
+        }
+
+        // use raw input mask channel
+        finalPage[blockOffset] = RunLengthEncodedBlock.create(BOOLEAN, null, positionCount);
+        return new Page(positionCount, finalPage);
+    }
+
+    public Page processRawInputPage(Page page)
+    {
+        Block[] finalPage = new Block[finalBlockCount];
+        int blockOffset = 0;
+        // raw input hash channels
+        for (int i = 0; i < hashChannels.length; i++, blockOffset++) {
+            finalPage[blockOffset] = page.getBlock(hashChannels[i]);
+        }
+        // aggregator state channels
+        for (int i = 0; i < groupedAggregators.size(); i++, blockOffset++) {
+            finalPage[blockOffset] = RunLengthEncodedBlock.create(groupedAggregators.get(i).getType(), null, page.getPositionCount());
```

Comment:
nit: storing type to just produce null RLE seems overkill, but I see it's being done like that throughout the code.